find_message_by_message_id still runs the #415 unindexed all-mailbox scan — freezes Mail, 2 integration tests red on main

`find_message_by_message_id` still runs the unindexed all-mailbox `whose message id` scan that #415 identified and removed everywhere else. It freezes Mail.app's UI and times out at 60s. Two integration tests are red on `main` because of it, and it froze a real Mail session during #421 work.

## Reproduced

Both of these fail on `main` (a1586ec), verified by stashing unrelated changes and re-running:

```
FAILED TestDraftsLifecycleIntegration::test_flag_two_ids_resolves_via_single_or_search
FAILED TestFindMessageByMessageIdIntegration::test_create_draft_reply_round_trip_with_bare_rfc_id

src/apple_mail_fast_mcp/mail_connector.py:4869: in find_message_by_message_id
    result = self._run_applescript(script).strip()
apple_mail_fast_mcp.exceptions.MailAppleScriptError: Script execution timeout after 60s
```

Mail.app was left wedged afterwards — a graceful `tell application "Mail" to quit` did not return within 15s and the process needed a SIGTERM.

## Mechanism

Two properties combine badly:

1. Mail does not index the `message id` property, so `whose message id is X` has no lookup path and must examine every message object.
2. AppleScript commands execute on Mail's main thread — the thread that draws the UI.

On a 5-account setup where Gmail's All Mail alone holds 61,880 messages, Mail walks tens of thousands of objects on its UI thread. The app stops redrawing and stops accepting input for the duration.

The 60s `_run_applescript` timeout does **not** rescue Mail. It kills our `osascript` client; Mail keeps grinding through work it already started. That is why the freeze outlives the test failure.

## Why it survived #415 and #413

- **#415** diagnosed this exact freeze and fixed `get_thread` by resolving RFC Message-IDs through indexed IMAP `SEARCH HEADER Message-ID` instead of AppleScript.
- **#413** narrowed the five draft-lifecycle loops to Mail's unified `drafts mailbox`. Its commit message notes the side win that "draft lookups no longer scan every mailbox, so they're fast on large Gmail accounts where the old broad scan hit the 60s osascript timeout."
- **`find_message_by_message_id` was deliberately left broad**, on the stated grounds that "reply/forward seed resolution needs to find seeds outside Drafts."

That reasoning is about **reach**, not **mechanism** — and the two got conflated. Needing to search widely does not require searching slowly. This is a gap left by a surgical fix, not a considered tradeoff.

## The fix already exists in this codebase

`SEARCH HEADER Message-ID` over those same 61,880 messages measures **0.14s**, server-side and indexed — see the [#420 measurements](https://github.com/s-morgan-jeffries/apple-mail-fast-mcp/issues/420#issuecomment-5158294014). `ImapConnector.resolve_anchor` already does exactly this lookup. `find_message_by_message_id` simply does not call it.

Direction:

- Route the lookup through indexed IMAP first, mirroring `_resolve_anchor_via_imap`.
- Keep AppleScript only as a **bounded** fallback — follow #419's pattern (bounded probe of the unified mailboxes) rather than the current unbounded accounts × mailboxes walk.
- Preserve the wide reach the #413 comment was protecting: the replacement must still resolve seeds outside Drafts. That is what IMAP's All-Mail / multi-folder search gives, without the UI-thread cost.

## Constraints

- Must not regress #425: a probe that *fails* is not the same as a message that is absent. Reuse the indeterminate/definitive distinction rather than collapsing errors into "not found".
- Must not regress #205 / #148: `create_draft(seed_id=<RFC Message-ID>)` depends on this resolution, and read tools emit bare RFC Message-IDs.
- Accounts without IMAP configured still need a working path — bounded AppleScript, not the current unbounded scan.

## Verification

The two failing integration tests above are the acceptance criteria; both must pass. Additionally, time a `find_message_by_message_id` call against the large Gmail account and confirm Mail stays interactive throughout (the current behavior is an outright UI freeze, so "does Mail respond while it runs" is a meaningful check, not a nicety).

**Caution when working this issue:** running the draft integration suite against a large account currently freezes Mail. Expect to restart it.

--- a/src/apple_mail_fast_mcp/imap_connector.py
+++ b/src/apple_mail_fast_mcp/imap_connector.py
@@ -1290,6 +1290,59 @@ class ImapConnector:
                 entry.get(b"BODYSTRUCTURE")
             )
 
+    def locate_message(self, message_id: str) -> dict[str, Any] | None:
+        """Locate an RFC 5322 Message-ID and return where and *when* it is.
+
+        Returns ``{"folder": str, "uid": int, "internaldate": datetime}`` for
+        the first probed folder containing the Message-ID, or ``None`` when no
+        probed folder has it.
+
+        The ``internaldate`` is the point of this method (#432). Mail.app's
+        ``message id`` property is UNINDEXED, so matching it in AppleScript
+        loads every message in a mailbox and freezes Mail's UI. Mail *does*
+        order messages newest-first with cheap positional access, so a caller
+        that knows roughly WHEN a message arrived can binary-search to it in
+        ~log2(n) property reads instead of scanning. This supplies that date
+        from the server side, where the lookup IS indexed (measured 0.14s over
+        61,880 messages).
+
+        Shares the bounded probe-folder set with :meth:`resolve_anchor` — for
+        locating, Gmail's All Mail is ideal rather than a liability, since it
+        mirrors every message and so answers for any folder.
+
+        Raises:
+            IMAPClientError / OSError / LoginError: connection/auth failures,
+                so the caller can distinguish "not there" from "could not
+                check" (#425).
+        """
+        bracketed = _bracket_message_id(message_id)
+        with self._session() as client:
+            for folder in self._anchor_probe_folders(client):
+                try:
+                    client.select_folder(folder, readonly=True)
+                    uids = client.search(["HEADER", "Message-ID", bracketed])
+                except IMAPClientError as exc:
+                    logger.debug(
+                        "locate_message: skipping %s (%s)", folder, exc
+                    )
+                    continue
+                if not uids:
+                    continue
+                try:
+                    fetched = client.fetch([uids[0]], [b"INTERNALDATE"])
+                except IMAPClientError:
+                    continue
+                entry = fetched.get(uids[0]) or {}
+                internaldate = entry.get(b"INTERNALDATE")
+                if internaldate is None:
+                    continue
+                return {
+                    "folder": folder,
+                    "uid": uids[0],
+                    "internaldate": internaldate,
+                }
+        return None
+
     def resolve_anchor(self, message_id: str) -> dict[str, Any] | None:
         """Resolve an RFC 5322 Message-ID to a get_thread anchor via
         server-side, INDEXED ``SEARCH HEADER Message-ID`` (#415).

--- a/src/apple_mail_fast_mcp/mail_connector.py
+++ b/src/apple_mail_fast_mcp/mail_connector.py
@@ -3449,11 +3449,18 @@ class AppleMailConnector:
         emits Mail's INDEXED integer ``id is N`` (#415/#416).
 
         Measured on a 33k-message Gmail account: 1.3s here vs 3.0s for the
-        full walk. (Handing the id to ``ImapConnector.resolve_anchor``
-        instead — the other obvious route to the same anchor — measured 17s
-        on that account, because ``SEARCH HEADER Message-ID`` over a 33k All
-        Mail is anything but instant. Hence AppleScript on both sides of this
-        branch.)
+        full walk.
+
+        (An earlier version of this note claimed that routing through
+        ``ImapConnector.resolve_anchor`` "measured 17s … ``SEARCH HEADER
+        Message-ID`` over a 33k All Mail is anything but instant". That
+        attribution was wrong and #420 corrected it: measured directly, the
+        SEARCH is **0.14s** over 61,880 messages — it is server-indexed. The
+        17s was orchestration, one AppleScript config resolve plus Keychain
+        read plus connect/login per account until the anchor's account came
+        up. AppleScript is still right for *this* branch, but because a
+        numeric id hits Mail's own indexed integer ``id``, not because IMAP
+        is slow. #432 depends on the corrected fact.)
 
         Returns:
             The anchor dict, or ``None`` when the id is in neither probed
@@ -4837,38 +4844,180 @@ class AppleMailConnector:
         """
         if not rfc5322_message_id:
             return None
-        bare = _bare_message_id(rfc5322_message_id)
-        bracketed = f"<{bare}>"
-        safe_bare = escape_applescript_string(sanitize_input(bare))
-        safe_bracketed = escape_applescript_string(sanitize_input(bracketed))
 
+        # Stage 1 (#432): ask IMAP WHERE and WHEN. Indexed server-side —
+        # measured 0.14s over 61,880 messages. Failures here are
+        # "could not determine", never "absent" (#425).
+        try:
+            located = self._locate_via_imap(rfc5322_message_id)
+        except _IMAP_FALLBACK_EXCS as exc:
+            raise MailAnchorLookupIncompleteError(
+                f"Could not determine whether Message-ID "
+                f"{rfc5322_message_id!r} exists: the IMAP lookup failed "
+                f"({type(exc).__name__}: {exc}). Mail.app's `message id` is "
+                "unindexed, so there is no cheap AppleScript fallback — "
+                "scanning for it freezes Mail (#432). Retry, or run "
+                "`apple-mail-fast-mcp setup-imap` if this account has no "
+                "IMAP credentials."
+            ) from exc
+        if located is None:
+            # A clean IMAP answer. The probe set is Gmail's All Mail (which
+            # mirrors every message) or INBOX+Sent — so this is definitive.
+            return None
+
+        # Stage 2: binary-search Mail by the date IMAP reported.
+        return self._find_by_message_id_near_date(
+            rfc5322_message_id, cast(_datetime, located["internaldate"])
+        )
+
+    def _locate_via_imap(self, rfc5322_message_id: str) -> dict[str, Any] | None:
+        """Ask each IMAP-configured account where/when a Message-ID lives.
+
+        Returns the first hit's ``{folder, uid, internaldate}``, or ``None``
+        when every configured account answered cleanly that it doesn't have
+        it. Raises on failure so the caller can distinguish "absent" from
+        "could not check" (#425).
+        """
+        last_exc: Exception | None = None
+        for acct in self.list_accounts():
+            account = cast(str, acct.get("name") or "")
+            if not account or self._imap_breaker_open(account):
+                continue
+            try:
+                host, port, email = self._resolve_imap_config(account)
+                if not host:
+                    continue
+                password = self._get_imap_password_with_fallback(account, email)
+                imap = ImapConnector(
+                    host, port, email, password, pool=self._imap_pool
+                )
+                found = imap.locate_message(rfc5322_message_id)
+            except MailKeychainEntryNotFoundError as exc:
+                # Benign opt-out, same as _resolve_anchor_via_imap (#425).
+                self._log_imap_fallback(account, exc)
+                continue
+            except _IMAP_FALLBACK_EXCS as exc:
+                self._log_imap_fallback(account, exc)
+                last_exc = exc
+                continue
+            if found is not None:
+                self._imap_clear_breaker(account)
+                return found
+        if last_exc is not None:
+            raise last_exc
+        return None
+
+    def _find_by_message_id_near_date(
+        self, rfc5322_message_id: str, when: _datetime
+    ) -> str | None:
+        """Binary-search Mail's unified mailboxes for a Message-ID known to
+        have arrived around ``when``, and return Mail's internal id (#432).
+
+        Mail orders messages strictly newest-first with cheap positional
+        access — verified on a 33,569-message INBOX: index 1 is 2026-08-09,
+        index 33,569 is 2004-11-12, and a read at index 33,568 is ~instant.
+        So the arrival date locates the message in ~log2(n) property reads
+        instead of the unindexed scan that freezes Mail.
+
+        The +/-1 day window absorbs the measured Mail-vs-IMAP date skew:
+        exact at indices 1/50/500/5000, but 1 hour adrift at index 20,000
+        (a DST/timezone artifact). A day of margin costs ~100 messages to
+        scan linearly and is cheap; a tight window would silently miss.
+        """
+        bare = _bare_message_id(rfc5322_message_id)
+        safe_bare = escape_applescript_string(sanitize_input(bare))
+        safe_bracketed = escape_applescript_string(
+            sanitize_input(f"<{bare}>")
+        )
+        # Build the date from COMPONENTS, never `date "..."`. AppleScript
+        # parses date-string literals against the user's locale: on this
+        # machine `date "2026-08-09 11:03:51"` yields "October 8, 12177",
+        # which silently makes any window comparison meaningless. Setting
+        # `day` to 1 first avoids overflow when the current day-of-month
+        # exceeds the target month's length.
+        secs_into_day = when.hour * 3600 + when.minute * 60 + when.second
         script = _wrap_with_timeout(
             f"""tell application "Mail"
+            set targetDate to (current date)
+            set day of targetDate to 1
+            set year of targetDate to {when.year}
+            set month of targetDate to {when.month}
+            set day of targetDate to {when.day}
+            set time of targetDate to {secs_into_day}
+            set loDate to targetDate - (1 * days)
+            set hiDate to targetDate + (1 * days)
             set foundId to ""
-            repeat with acc in accounts
-                try
-                    repeat with mb in mailboxes of acc
-                        try
-                            set m to first message of mb whose (message id is "{safe_bare}" or message id is "{safe_bracketed}")
-                            set foundId to (id of m as text)
-                            exit repeat
-                        end try
-                    end repeat
-                end try
+            set sawOrderProblem to false
+
+            repeat with mb in {{inbox, sent mailbox}}
+                set n to (count of messages of mb)
+                if n > 0 then
+                    -- #242: Mail's iteration order has changed before. A
+                    -- silent reversal would make the search below return
+                    -- wrong answers, so verify descending order first.
+                    if n > 1 then
+                        set firstDate to (date received of (message 1 of mb))
+                        set lastDate to (date received of (message n of mb))
+                        if firstDate < lastDate then set sawOrderProblem to true
+                    end if
+
+                    if not sawOrderProblem then
+                        -- First index whose date is at or below the window
+                        -- top (dates DESCEND as the index grows).
+                        set lo to 1
+                        set hi to n
+                        repeat while lo < hi
+                            set midIdx to (lo + hi) div 2
+                            if (date received of (message midIdx of mb)) > hiDate then
+                                set lo to midIdx + 1
+                            else
+                                set hi to midIdx
+                            end if
+                        end repeat
+
+                        -- Linear scan of the window only.
+                        repeat with i from lo to n
+                            set m to message i of mb
+                            if (date received of m) < loDate then exit repeat
+                            set mid to ""
+                            try
+                                set mid to message id of m
+                            end try
+                            if mid is "{safe_bare}" or mid is "{safe_bracketed}" then
+                                set foundId to (id of m as text)
+                                exit repeat
+                            end if
+                        end repeat
+                    end if
+                end if
                 if foundId is not "" then exit repeat
+                if sawOrderProblem then exit repeat
             end repeat
-            if foundId is "" then
-                return "NOT_FOUND"
-            else
-                return foundId
-            end if
+
+            if sawOrderProblem then return "ORDER_UNEXPECTED"
+            if foundId is "" then return "WINDOW_MISS"
+            return foundId
         end tell""",
             timeout=self.timeout,
         )
 
         result = self._run_applescript(script).strip()
-        if result == "NOT_FOUND" or not result:
-            return None
+        if result == "ORDER_UNEXPECTED":
+            raise MailAnchorLookupIncompleteError(
+                f"Could not locate Message-ID {rfc5322_message_id!r}: Mail is "
+                "not enumerating messages newest-first, so the indexed "
+                "date-bounded search is unsafe. Refusing to fall back to the "
+                "unindexed scan, which freezes Mail (#432/#242)."
+            )
+        if result == "WINDOW_MISS" or not result:
+            raise MailAnchorLookupIncompleteError(
+                f"Could not locate Message-ID {rfc5322_message_id!r} in Mail. "
+                f"IMAP reports it arrived {when:%Y-%m-%d %H:%M:%S}, but it is "
+                "not in the unified inbox or sent mailbox within a day of "
+                "that. It may live in another mailbox; searching every "
+                "mailbox by Message-ID is what freezes Mail, so that is not "
+                "attempted (#432)."
+            )
         return result
 
     def get_draft_state(self, draft_id: str) -> dict[str, Any]:
@@ -5080,19 +5229,30 @@ class AppleMailConnector:
             verb = "reply to all" if reply_all else "reply"
         else:  # forward
             verb = "forward"
+        # #432: this used to walk `accounts x mailboxes`, re-finding a message
+        # its caller had just resolved. Mail's `id` is an INDEXED integer, so
+        # probing the two unified mailboxes (locale-independent aggregations
+        # across every account, per #407/#419) is both correct and instant —
+        # and it is where find_message_by_message_id searched, so a seed it
+        # resolved is reachable here.
+        #
+        # The id is emitted UNQUOTED when numeric: `id` is an integer, and
+        # comparing it against a quoted string is what the old form did.
+        id_literal = (
+            seed_id_safe
+            if seed_id_safe and seed_id_safe.isdigit()
+            else f'"{seed_id_safe}"'
+        )
         return f"""
             set origMsg to missing value
-            repeat with acc in accounts
+            try
+                set origMsg to first message of inbox whose id is {id_literal}
+            end try
+            if origMsg is missing value then
                 try
-                    repeat with mb in mailboxes of acc
-                        try
-                            set origMsg to first message of mb whose id is "{seed_id_safe}"
-                            exit repeat
-                        end try
-                    end repeat
+                    set origMsg to first message of sent mailbox whose id is {id_literal}
                 end try
-                if origMsg is not missing value then exit repeat
-            end repeat
+            end if
             if origMsg is missing value then error "SEED_NOT_FOUND"
             set theMessage to {verb} origMsg opening window false
         """

--- a/tests/integration/test_mail_integration.py
+++ b/tests/integration/test_mail_integration.py
@@ -2798,6 +2798,66 @@ class TestFindMessageByMessageIdIntegration:
     survives any specific Message-ID being deleted.
     """
 
+    def test_mail_stays_responsive_during_lookup(
+        self, connector: AppleMailConnector, test_account: str
+    ) -> None:
+        """#432: the point of this fix is that Mail keeps drawing.
+
+        The old `whose message id` walk ran on Mail's UI thread across every
+        mailbox of every account — measured here at 33,569 messages in Gmail's
+        INBOX and 62,085 in All Mail — so Mail stopped responding entirely and
+        stayed wedged after the 60s timeout killed our osascript client.
+
+        This asserts the observable symptom, not a proxy for it: a second,
+        trivial AppleScript must get an answer WHILE the lookup runs. Against
+        the pre-fix code it does not.
+        """
+        import subprocess
+        import threading
+
+        rows = connector.search_messages(
+            account=test_account, mailbox="INBOX", limit=1
+        )
+        if not rows:
+            pytest.skip(f"{test_account} INBOX has no messages")
+        rfc_id = rows[0].get("rfc_message_id") or rows[0].get("id")
+        if not rfc_id or "@" not in rfc_id:
+            pytest.skip("test_account is not on the IMAP path")
+
+        done = threading.Event()
+
+        def _lookup() -> None:
+            try:
+                connector.find_message_by_message_id(str(rfc_id))
+            finally:
+                done.set()
+
+        t = threading.Thread(target=_lookup, daemon=True)
+        t.start()
+        try:
+            probe_start = time.monotonic()
+            probe = subprocess.run(
+                ["/usr/bin/osascript", "-e",
+                 'tell application "Mail" to return (count of accounts) as text'],
+                capture_output=True, text=True, timeout=30,
+            )
+            probe_elapsed = time.monotonic() - probe_start
+        except subprocess.TimeoutExpired:
+            pytest.fail(
+                "Mail did not answer a trivial AppleScript within 30s while "
+                "find_message_by_message_id was running — the UI thread is "
+                "blocked, i.e. the #432 freeze is back"
+            )
+        finally:
+            done.wait(timeout=120)
+        t.join(timeout=5)
+
+        assert probe.returncode == 0, f"probe failed: {probe.stderr}"
+        assert probe_elapsed < 25.0, (
+            f"Mail took {probe_elapsed:.1f}s to answer `count of accounts` "
+            "during the lookup — the UI thread is being starved (#432)"
+        )
+
     def test_find_by_bare_rfc_id_from_search_messages(
         self, connector: AppleMailConnector, test_account: str
     ) -> None:

--- a/tests/unit/test_mail_connector.py
+++ b/tests/unit/test_mail_connector.py
@@ -4,6 +4,7 @@ import logging
 import smtplib
 import time
 import warnings
+from datetime import datetime
 from pathlib import Path
 from typing import Any
 from unittest.mock import MagicMock, patch
@@ -5263,6 +5264,13 @@ class TestNonJsonPathsThreadTimeout:
     def test_find_message_by_message_id(self, mock_run: MagicMock) -> None:
         connector = AppleMailConnector(timeout=122)
         mock_run.return_value = MagicMock(returncode=0, stdout="42", stderr="")
+        # #432: the RFC path now consults IMAP for the arrival date before
+        # touching AppleScript. Stub that so this stays a test about the
+        # timeout clause on the AppleScript it then emits.
+        connector._locate_via_imap = lambda mid: {  # type: ignore[assignment]
+            "folder": "INBOX",
+            "internaldate": datetime(2026, 8, 9, 11, 3, 51),
+        }
         connector.find_message_by_message_id("<abc@example.com>")
         assert "with timeout of 122 seconds" in self._script_from(mock_run)
 
@@ -5695,7 +5703,114 @@ class TestFindMessageByMessageId:
 
     @pytest.fixture
     def connector(self) -> AppleMailConnector:
-        return AppleMailConnector(timeout=30)
+        c = AppleMailConnector(timeout=30)
+        # #432: the RFC path is IMAP-assisted. Default every test in this
+        # class to "IMAP located it on 2026-08-09" unless it says otherwise.
+        c._locate_via_imap = lambda mid: {  # type: ignore[assignment]
+            "folder": "[Gmail]/All Mail",
+            "internaldate": datetime(2026, 8, 9, 11, 3, 51),
+        }
+        return c
+
+    # --- #432: no unindexed all-mailbox scan, ever -----------------------
+
+    @patch.object(AppleMailConnector, "_run_applescript")
+    def test_script_never_walks_accounts_x_mailboxes(
+        self, mock_run: MagicMock, connector: AppleMailConnector
+    ) -> None:
+        """The freeze: `whose message id` is UNINDEXED, so pairing it with an
+        accounts x mailboxes walk loads tens of thousands of messages on
+        Mail's UI thread. Measured: Gmail INBOX 33,569 / All Mail 62,085."""
+        mock_run.return_value = "160989"
+        connector.find_message_by_message_id("<abc@example.com>")
+        script = mock_run.call_args[0][0]
+        assert "repeat with acc in accounts" not in script
+        assert "mailboxes of acc" not in script
+
+    @patch.object(AppleMailConnector, "_run_applescript")
+    def test_script_binary_searches_by_index(
+        self, mock_run: MagicMock, connector: AppleMailConnector
+    ) -> None:
+        """Messages are strictly newest-first and positional access is cheap
+        (verified: idx 1 = 2026-08-09, idx 33,569 = 2004-11-12; a read at
+        33,568 is ~instant). So the date from IMAP locates the message in
+        ~log2(n) property reads instead of a scan."""
+        mock_run.return_value = "160989"
+        connector.find_message_by_message_id("<abc@example.com>")
+        script = mock_run.call_args[0][0]
+        assert "div 2" in script, "expected a binary search"
+        assert "date received" in script
+
+    @patch.object(AppleMailConnector, "_run_applescript")
+    def test_script_guards_the_newest_first_assumption(
+        self, mock_run: MagicMock, connector: AppleMailConnector
+    ) -> None:
+        """#242 records that Mail's iteration order CHANGED once already. A
+        silent reversal would make the binary search return wrong answers, so
+        the script must verify descending order before trusting it."""
+        mock_run.return_value = "160989"
+        connector.find_message_by_message_id("<abc@example.com>")
+        script = mock_run.call_args[0][0]
+        assert "ORDER_UNEXPECTED" in script
+
+    @patch.object(AppleMailConnector, "_run_applescript")
+    def test_date_is_built_from_components_never_a_string_literal(
+        self, mock_run: MagicMock, connector: AppleMailConnector
+    ) -> None:
+        """AppleScript parses `date "..."` against the user's LOCALE. On a real
+        machine `date "2026-08-09 11:03:51"` evaluated to "October 8, 12177" —
+        a 10,000-year error that silently made the window match nothing, with
+        no error raised. Component assignment is locale-independent."""
+        mock_run.return_value = "160989"
+        connector.find_message_by_message_id("<abc@example.com>")
+        script = mock_run.call_args[0][0]
+        assert 'date "' not in script, "locale-dependent date literal"
+        assert "set year of targetDate to 2026" in script
+        assert "set month of targetDate to 8" in script
+        assert "set day of targetDate to 9" in script
+        # `day` reset to 1 first, or setting month can overflow the day.
+        assert script.index("set day of targetDate to 1") < script.index(
+            "set month of targetDate to 8"
+        )
+
+    def test_imap_failure_is_indeterminate_not_absent(
+        self, connector: AppleMailConnector
+    ) -> None:
+        """#425's lesson: a lookup that could not run is not a message that
+        does not exist. Returning None here would make callers report
+        'no such message' for a message that is simply unreachable."""
+        def _boom(mid: str) -> None:
+            raise OSError("cannot read from timed out object")
+
+        connector._locate_via_imap = _boom  # type: ignore[assignment]
+        with pytest.raises(MailAnchorLookupIncompleteError):
+            connector.find_message_by_message_id("<abc@example.com>")
+
+    @patch.object(AppleMailConnector, "_run_applescript")
+    def test_window_miss_is_indeterminate_not_absent(
+        self, mock_run: MagicMock, connector: AppleMailConnector
+    ) -> None:
+        """IMAP proved the message exists; the date-bounded window failed to
+        find it in Mail. That is 'could not determine', not 'absent'."""
+        mock_run.return_value = "WINDOW_MISS"
+        with pytest.raises(MailAnchorLookupIncompleteError):
+            connector.find_message_by_message_id("<abc@example.com>")
+
+    @patch.object(AppleMailConnector, "_run_applescript")
+    def test_ordering_guard_trip_is_indeterminate(
+        self, mock_run: MagicMock, connector: AppleMailConnector
+    ) -> None:
+        mock_run.return_value = "ORDER_UNEXPECTED"
+        with pytest.raises(MailAnchorLookupIncompleteError):
+            connector.find_message_by_message_id("<abc@example.com>")
+
+    def test_imap_says_absent_returns_none(
+        self, connector: AppleMailConnector
+    ) -> None:
+        """A clean IMAP answer of 'not in any probed folder' IS definitive —
+        All Mail mirrors every message — so None (absent) is correct here."""
+        connector._locate_via_imap = lambda mid: None  # type: ignore[assignment]
+        assert connector.find_message_by_message_id("<gone@example.com>") is None
 
     @patch.object(AppleMailConnector, "_run_applescript")
     def test_returns_internal_id_on_match(
@@ -5707,11 +5822,13 @@ class TestFindMessageByMessageId:
         )
         assert result == "160989"
 
-    @patch.object(AppleMailConnector, "_run_applescript")
     def test_returns_none_on_not_found(
-        self, mock_run: MagicMock, connector: AppleMailConnector
+        self, connector: AppleMailConnector
     ) -> None:
-        mock_run.return_value = "NOT_FOUND"
+        """#432 moved the absence decision to IMAP, which answers definitively
+        (Gmail's All Mail mirrors every message). AppleScript no longer emits
+        a NOT_FOUND sentinel — it is never asked to search blind."""
+        connector._locate_via_imap = lambda mid: None  # type: ignore[assignment]
         result = connector.find_message_by_message_id(
             "<missing@example.com>"
         )
@@ -5740,22 +5857,26 @@ class TestFindMessageByMessageId:
         """Quotes/backslashes in the Message-ID must be escaped to prevent
         AppleScript injection. Real Message-IDs almost never contain these
         but we shouldn't trust the wire."""
-        mock_run.return_value = "NOT_FOUND"
+        mock_run.return_value = "160989"
         connector.find_message_by_message_id('<weird"id@host>')
         script = mock_run.call_args[0][0]
         # Escaped quote inside the AppleScript string literal.
         assert '\\"' in script
 
     @patch.object(AppleMailConnector, "_run_applescript")
-    def test_script_uses_whose_message_id_clause(
+    def test_script_matches_message_id_without_a_whose_clause(
         self, mock_run: MagicMock, connector: AppleMailConnector
     ) -> None:
-        mock_run.return_value = "NOT_FOUND"
+        """#432 inverts the old assertion. `whose message id is ...` is the
+        UNINDEXED filter that makes Mail load every message in the mailbox;
+        the match now happens per-message inside the date-bounded window,
+        where the candidate set is ~a day of mail."""
+        mock_run.return_value = "160989"
         connector.find_message_by_message_id("<x@y>")
         script = mock_run.call_args[0][0]
-        # Compound clause queries both bare and bracketed forms (#205 follow-up).
-        assert "whose" in script
-        assert "message id is" in script
+        assert "whose message id" not in script
+        assert "whose (message id" not in script
+        assert "message id of m" in script
 
     @patch.object(AppleMailConnector, "_run_applescript")
     def test_bracketless_input_queries_both_forms(
@@ -5767,12 +5888,12 @@ class TestFindMessageByMessageId:
         resolver therefore queries both forms in a single ``whose``
         clause so a caller doesn't need to know the storage convention.
         """
-        mock_run.return_value = "NOT_FOUND"
+        mock_run.return_value = "160989"
         connector.find_message_by_message_id("abc@example.com")
         script = mock_run.call_args[0][0]
-        assert 'message id is "abc@example.com"' in script
-        assert 'message id is "<abc@example.com>"' in script
-        assert "whose" in script and " or " in script
+        assert 'mid is "abc@example.com"' in script
+        assert 'mid is "<abc@example.com>"' in script
+        assert " or " in script
 
     @patch.object(AppleMailConnector, "_run_applescript")
     def test_bracketed_input_queries_both_forms(
@@ -5782,11 +5903,11 @@ class TestFindMessageByMessageId:
         already include brackets; strip them and query both forms so
         we don't depend on Mail.app's storage convention.
         """
-        mock_run.return_value = "NOT_FOUND"
+        mock_run.return_value = "160989"
         connector.find_message_by_message_id("<abc@example.com>")
         script = mock_run.call_args[0][0]
-        assert 'message id is "abc@example.com"' in script
-        assert 'message id is "<abc@example.com>"' in script
+        assert 'mid is "abc@example.com"' in script
+        assert 'mid is "<abc@example.com>"' in script
         assert "<<" not in script and ">>" not in script
 
     @patch.object(AppleMailConnector, "_run_applescript")
@@ -6107,6 +6228,25 @@ class TestCreateDraft:
         assert "tell theMessage to send" in script
         # No diff snapshot when sending.
         assert "set beforeIds to" not in script
+
+    @patch.object(AppleMailConnector, "_run_applescript")
+    def test_reply_seed_lookup_never_walks_accounts_x_mailboxes(
+        self, mock_run: MagicMock, connector: AppleMailConnector
+    ) -> None:
+        """#432: the creation block used to re-walk every mailbox of every
+        account to find the seed its caller had *just* resolved. Fixing only
+        find_message_by_message_id would have left this second walk in place,
+        so the freeze would have survived.
+
+        Mail's `id` is an indexed integer, so the two unified mailboxes
+        answer instantly."""
+        mock_run.return_value = "160989"
+        connector.create_draft(seed="reply", seed_id="12345", body="x")
+        script = mock_run.call_args[0][0]
+        assert "repeat with acc in accounts" not in script
+        assert "mailboxes of acc" not in script
+        assert "first message of inbox whose id is 12345" in script
+        assert "first message of sent mailbox whose id is 12345" in script
 
     # --- #421: the id-bridging diff must wait for the draft to appear ----
 
@@ -6453,7 +6593,9 @@ class TestCreateDraft:
         mock_resolve.assert_called_once_with("abc-123@example.com")
         script = mock_run.call_args[0][0]
         # AppleScript looks up by Mail's internal id, not the RFC id.
-        assert 'whose id is "160989"' in script
+        # #432: the id is an INDEXED integer property, so it is emitted
+        # unquoted now; the old quoted form compared int to string.
+        assert 'whose id is 160989' in script
         assert "abc-123@example.com" not in script
 
     @patch.object(AppleMailConnector, "find_message_by_message_id")
@@ -6475,7 +6617,9 @@ class TestCreateDraft:
         )
         mock_resolve.assert_called_once_with("abc-123@example.com")
         script = mock_run.call_args[0][0]
-        assert 'whose id is "160989"' in script
+        # #432: the id is an INDEXED integer property, so it is emitted
+        # unquoted now; the old quoted form compared int to string.
+        assert 'whose id is 160989' in script
 
     @patch.object(AppleMailConnector, "find_message_by_message_id")
     @patch.object(AppleMailConnector, "_run_applescript")
@@ -6496,7 +6640,9 @@ class TestCreateDraft:
         )
         mock_resolve.assert_not_called()
         script = mock_run.call_args[0][0]
-        assert 'whose id is "160989"' in script
+        # #432: the id is an INDEXED integer property, so it is emitted
+        # unquoted now; the old quoted form compared int to string.
+        assert 'whose id is 160989' in script
 
     # No from_account → create_draft would auto-resolve a sole account
     # (#321) via list_accounts; stub it out so this test isolates the RFC

